### PR TITLE
NextUp & NextDown

### PR DESCRIPTION
**Pull request by eggrobin**
We don’t have `std::bit_cast` in our *nix builds yet, and `absl::bit_cast` is not constexpr so it does not help.
Implement an IEEE 754 `nextUp` and `nextDown` the slow and ugly way, using arithmetic.
We will probably want to keep those even when we get C++20 everywhere, since `std::nextafter` does not have IEEE 754 semantics and is not constexpr anyway. We will want to reimplement them with `std::bit_cast` though.

### CODE REVIEW

File: astronomy/time_scales_body.hpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include "geometry/named_quantities.hpp"
 #include "glog/logging.h"
 #include "numerics/double_precision.hpp"
+#include "quantities/elementary_functions.hpp"
```
Comment (pleroy) [inline astronomy/time_scales_body.hpp:15]: Why this include?  Shouldn't you get `next.hpp`?

File: numerics/log_b.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace principia {
+namespace numerics {
+
+// A constexpr implementation of the IEEE 754:2008 logB function.
+// Uses sourceFormat as logBFormat, which makes it easy to cleanly handle NaN,
+// infinity, and 0.
+template<typename SourceFormat,
+         typename = std::enable_if_t<std::is_floating_point_v<SourceFormat>>>
+constexpr SourceFormat LogB(SourceFormat const x) {
```
Comment (pleroy) [inline numerics/log_b.hpp:13]: Please separate the `_body.hpp` file from the `.hpp` file as we normally do.

File: numerics/log_b.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include <limits>
+
+namespace principia {
+namespace numerics {
+
+// A constexpr implementation of the IEEE 754:2008 logB function.
+// Uses sourceFormat as logBFormat, which makes it easy to cleanly handle NaN,
```
Comment (pleroy) [inline numerics/log_b.hpp:9]: Weird camelCase (2x).

File: numerics/log_b.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  if (x == 0) {
+    // Signal the divideByZero exception (by dividing by -0).
+    return 1 / -abs_x;
+  }
+  SourceFormat scaled_abs_x = abs_x;
+  SourceFormat result = 0;
+  while (scaled_abs_x < 1) {
+    scaled_abs_x *= std::numeric_limits<SourceFormat>::radix;
+    result -= 1;
```
Comment (pleroy) [inline numerics/log_b.hpp:29]: `--result`

File: numerics/log_b.hpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  while (scaled_abs_x >= std::numeric_limits<SourceFormat>::radix) {
+    scaled_abs_x /= std::numeric_limits<SourceFormat>::radix;
+    result += 1;
```
Comment (pleroy) [inline numerics/log_b.hpp:33]: `++result`

File: numerics/next.hpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#include "numerics/scale_b.hpp"
+
+namespace principia {
+namespace numerics {
+
+// A constexpr implementation of the IEEE 754:2008 nextUp and nextDown
+// functions.
+template<typename SourceFormat,
+         typename = std::enable_if_t<std::is_floating_point_v<SourceFormat>>>
+constexpr SourceFormat NextUp(SourceFormat const x) {
```
Comment (pleroy) [inline numerics/next.hpp:16]: Separate the body.

File: numerics/next.hpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    return -0.0;
+  }
+  if (x == 0) {
+    // Note that this is independent of the sign of 0, as specified by
+    // IEEE 754:2008, 5.3.1.  This differs from C++ std::nextafter, which maps
+    // -0 to +0.
+    return std::numeric_limits<SourceFormat>::denorm_min();
+  }
+  if (x == std::numeric_limits<SourceFormat>::infinity()) {
+    // TODO(egg): Should we be signalling the overflow exception?
```
Comment (pleroy) [inline numerics/next.hpp:29]: I don't think we ever use the floating-point exceptions...

File: numerics/next_test.cpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+using geometry::Sign;
+using ::testing::Eq;
+
+class NextTest : public ::testing::Test {};
+
+TEST_F(NextTest, ExtendedRealLine) {
+  static_assert(NextUp(-std::numeric_limits<double>::infinity()) ==
+                -0x1.FFFFFFFFFFFFFp+1023);
+  static_assert(NextUp(-0x1.FFFFFFFFFFFFFp+1023) == -0x1.FFFFFFFFFFFFEp+1023);
+  // [...]
```
Comment (pleroy) [inline numerics/next_test.cpp:21]: No idea what this comment is trying to tell me.

**Review by pleroy — COMMENTED** [review]
For `scale_b` too, we should separate the body.

File: numerics/log_b.hpp (hunk 3 above)
Comment (eggrobin) [reply numerics/log_b.hpp:9]: IEEE 754 uses that case, and I am referring to names in there.

File: numerics/next_test.cpp (hunk 8 above)
Comment (eggrobin) [reply numerics/next_test.cpp:21]: A few numbers were omitted in the progression from -∞ to +∞.

File: numerics/next.hpp (hunk 7 above)
Comment (eggrobin) [reply numerics/next.hpp:29]: No, but when it’s trivial to get them right we might as well…

File: astronomy/time_scales_body.hpp (hunk 1 above)
Comment (eggrobin) [reply astronomy/time_scales_body.hpp:15]: `numerics/next.hpp` only has the one for `double`.

**Pull request merged**